### ISSUE TITLE
PersistentStreamProviders cause grain activations before Silo initialization is complete

### ISSUE THREAD

**Issue by randa1** [dotnet/orleans#1110]
In `PersistentStreamPullingAgent`'s `Initialize` method the message consuming timer is immediately started with a random time interval (line 156):
``` c#
// Setup a reader for a new receiver. 
// Even if the receiver failed to initialise, treat it as OK and start pumping it. It's receiver responsibility to retry initialization.
var randomTimerOffset = safeRandom.NextTimeSpan(config.GetQueueMsgsTimerPeriod);
timer = base.RegisterTimer(AsyncTimerCallback, QueueId, randomTimerOffset, config.GetQueueMsgsTimerPeriod);
```

Once the timer ticks, messages will be read from the physical queue and delivered to consumer grains - thus causing their activation, loading of state, etc. In explicit streams, it will also load the `PubSubRendevouzGrain`s. 

Since stream providers are loaded before bootstrap providers, this causes a race condition. All of this happens before the Silo bootstrap providers are loaded and before Silo initialization is complete. In our application, there's code in the bootstrap provider which must happen before the initialization of the grains, so this results in our grains being loaded with a "bad" state. 

I think this should be handled differently, and this is my suggestion:
1. Add a new method to the `IProvider` (or specifically to `IStreamProvider`) interface - `OnSiloLoaded`. 
2. Call `OnSiloLoaded` for all providers after the Silo initialization is complete.
3. The code above from the pulling agent will be moved to `OnSiloLoaded`.

What do you think?

**Comment by gabikliot** [comment]
I agree in general that we need to start the timer later on.
I want to look how to better add this new initialization phase, since we already have Load, Init and Start. Might be a way to not add a 4th phase.

BTW - I think there is a bigger issue here.
You wrote: "there's code in the bootstrap provider which must happen before the initialization of the grains". Even if we solve the stream provider init problem you raised, the grain can be activated on a silo BEFORE bootstrap provider runs. The order is: load storage providers (they are ready to be used after loaded, they do not need to be started), then Silo Becomes Active in Membership - after that phase OTHER silos may send msg to this silo and activate grain on it. So from a distributed perspective this silo is fully started and functional once it become Active in Membership. And only later we start Bootstrap.

Now you may suggest to start Bootstrap earlier, before became Active. But that would mean that the Bootstrap provider cannot send msgs to grains, since a not-yet-active silo cannot send msg to grains. It definitely cannot send msgs to other silos, they will reject it and locally now we also consider only Active Silo to be a fully functional silo.

Maybe one solution here: maybe we need 2 types of Bootstrap providers - one started before silo starts and the other after. The first type can not send msg to grain (we don't inject grain factory into it so it cannot even create grain references).
That will solve your problem with stream provider as well, since stream provider in your case is simply yet another source of msg, just like any other silo.

Thoughts? @sergeybykov , @jason-bragg ?

**Comment by gabikliot** [comment]
So to be clear: what I am suggesting, instead of splitting the stream init sequence, extend the bootstrap provider in not 2 parts. Either one bootstrap provider with 2 calls: `InitBeforeSiloStarted `and `InitAfterSiloStarted` or 2 separate bootstrap providers (Before and After). The InitBeforeSiloStarted is not allowed to send msgs to grains or use any other runtime services (such as subscribe to streams or to reminders). It can interact with other providers, like storage provider for example, via direct reference. InitAfterSiloStarted can so anything.

@randa1 would that solve your scenario?

**Comment by randa1** [comment]
@gabikliot Yes, that would solve this specific scenario.

But I still think that the `PersistentStreamProvider`'s timer should be started after Silo is finished initialization, because it the stream provider can load a lot of messages from the queue, and starting a lot of grains at once has a significant effect on CPU and IO - which may slow down the initialization process.

**Comment by gabikliot** [comment]
Fixed via #1116.
I still think we will need to eventually separate Boostrap provider into 2 calls: before silo started and after, but for now this seems enough to solve your case.
